src/binary-reader.cc

```diff
@@ -2028,7 +2039,7 @@ Result read_binary(const void* data,
   WABT_ZERO_MEMORY(logging_reader);
   logging_reader.user_data = &logging_context;
 
-  logging_reader.on_error = reader->on_error;
+  logging_reader.on_error = logging_on_error;
```

Comment:
The reason I did this is because some tools don't register an on_error handler and what to default to default one.   If logging is enabled the default handler is not used, so the error gets swallowed.  Try wasmdump or wasm-link with --debug on a file with an error to see this.